moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp

```diff
@@ -874,83 +880,342 @@ bool TimeOptimalTrajectoryGeneration::computeTimeStamps(robot_trajectory::RobotT
+    {
+      ROS_ERROR_STREAM_NAMED(LOGNAME, "No acceleration limit was defined for joint "
+                                          << vars[idx].c_str()
+                                          << "! You have to define acceleration limits in the URDF or "
+                                             "joint_limits.yaml");
+      return false;
     }
   }
 
+  return doTimeParameterizationCalculations(trajectory, max_velocity, max_acceleration);
+}
+
+bool TimeOptimalTrajectoryGeneration::computeTimeStampsWithTorqueLimits(
+    robot_trajectory::RobotTrajectory& trajectory, const geometry_msgs::Vector3& gravity_vector,
+    const std::vector<double>& joint_torque_limits, double accel_limit_decrement_factor,
+    const std::unordered_map<std::string, double>& velocity_limits,
+    const std::unordered_map<std::string, double>& acceleration_limits, const double max_velocity_scaling_factor,
+    const double max_acceleration_scaling_factor) const
+{
+  // 1. Call computeTimeStamps() to time-parameterize the trajectory with given vel/accel limits.
+  // 2. Run forward dynamics to check if torque limits are violated at any waypoint.
+  // 3. If a torque limit was violated, decrease the acceleration limit for that joint and go back to Step 1.
+
+  if (trajectory.empty())
+    return true;
+
+  const moveit::core::JointModelGroup* group = trajectory.getGroup();
+  if (!group)
+  {
+    ROS_ERROR_NAMED(LOGNAME, "It looks like the planner did not set the group the plan was computed for");
+    return false;
+  }
+
+  // Create a mutable copy of acceleration_limits
+  std::unordered_map<std::string, double> mutable_accel_limits = acceleration_limits;
+
+  size_t dof = group->getActiveJointModels().size();
+
+  if (joint_torque_limits.size() != dof)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Joint torque limit vector size (" << joint_torque_limits.size()
+                                                                       << ") does not match the DOF of the RobotModel ("
+                                                                       << dof << ")");
+    return false;
+  }
+
+  if (accel_limit_decrement_factor < 0.01 || accel_limit_decrement_factor > 0.2)
+  {
+    ROS_ERROR_NAMED(LOGNAME, "The accel_limit_decrement_factor is outside the typical range [0.01, 0.2]");
+    return false;
+  }
+
+  dynamics_solver::DynamicsSolver dynamics_solver(trajectory.getRobotModel(), group->getName(), gravity_vector);
+
+  // Assume no external forces on the robot. This could easily be an argument later.
+  const std::vector<geometry_msgs::Wrench> link_wrenches = [&group] {
+    geometry_msgs::Wrench zero_wrench;
+    zero_wrench.force.x = 0;
+    zero_wrench.force.y = 0;
+    zero_wrench.force.z = 0;
+    zero_wrench.torque.x = 0;
+    zero_wrench.torque.y = 0;
+    zero_wrench.torque.z = 0;
+    // KDL (the dynamics solver) requires one wrench per link
+    std::vector<geometry_msgs::Wrench> vector_of_zero_wrenches(group->getLinkModels().size());
```

Comment:
```suggestion
    std::vector<geometry_msgs::Wrench> vector_of_zero_wrenches(group->getLinkModels().size(), zero_wrench);
```
looks like you wanted to do this?